drivers/dac/dac_mcp4725.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2021 Laird Connectivity
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT microchip_mcp4725
+
+#include <zephyr.h>
+#include <drivers/i2c.h>
+#include <drivers/dac.h>
+#include <logging/log.h>
+#include <dt-bindings/dac/mcp4725.h>
+
+LOG_MODULE_REGISTER(dac_mcp4725, CONFIG_DAC_LOG_LEVEL);
+
+/* Information in this file comes from MCP4725 datasheet revision D
+ * found at https://ww1.microchip.com/downloads/en/DeviceDoc/22039d.pdf
+ */
+
+/* Bits that can't be set when an application wants to write to dac */
+#define MCP4725_VALUE_DONT_SET_BITS			0x07FFF000
+
+/* Defines for field values in MCP4725 DAC register */
+#define MCP4725_WRITE_MODE_Mask				0x7
+#define MCP4725_POWER_DOWN_Mask				0x3
+#define MCP4725_DAC_VAL_Mask                            0xFFF
+
+#define MCP4725_FAST_MODE_POWER_DOWN_Pos		4
+#define MCP4725_FAST_MODE_DAC_UPPER_VAL_Pos		8
+#define MCP4725_FAST_MODE_DAC_UPPER_VAL_Mask		0xF
+#define MCP4725_FAST_MODE_DAC_LOWER_VAL_Mask		0xFF
+
+#define MCP4725_WRITE_DAC_MODE_WRITE_MODE_Pos		5
+#define MCP4725_WRITE_DAC_MODE_POWER_DOWN_Pos		1
+#define MCP4725_WRITE_DAC_MODE_DAC_UPPER_VAL_Pos	4
+#define MCP4725_WRITE_DAC_MODE_DAC_UPPER_VAL_Mask	0xFF
+#define MCP4725_WRITE_DAC_MODE_DAC_LOWER_VAL_Pos	4
+#define MCP4725_WRITE_DAC_MODE_DAC_LOWER_VAL_Mask	0xF0
+
+#define MCP4725_READ_RDY_Pos				7
+#define MCP4725_READ_RDY_Mask				0x1
+
+/* After writing eeprom, the MCP4725 can be in a busy state for 25 - 50ms
+ * See section 1.0 of MCP4725 datasheet, 'Electrical Characteristics'
+ */
+#define MCP4725_BUSY_TIMEOUT_MS                         60
+
+struct mcp4725_config {
+	const char *i2c_master_dev_name;
+	uint16_t i2c_slave_addr;
+};
+
+struct mcp4725_data {
+	const struct device *i2c_master;
+};
+
+/* Read mcp4725 and check RDY status bit */
+static int mcp4725_wait_until_ready(const struct device *dev, uint16_t slave_addr)
+{
+	uint8_t rxData[5];
+	bool mcp4725_ready = false;
+	int32_t retVal;
+	int32_t timeout = k_uptime_get_32() + MCP4725_BUSY_TIMEOUT_MS;
+
+	/* Wait until RDY bit is set or return error if timer exceeds MCP4725_BUSY_TIMEOUT_MS */
+	while (!mcp4725_ready) {
+		retVal = i2c_read(dev, rxData, sizeof(rxData), slave_addr);
+
+		if (retVal == 0) {
+			mcp4725_ready = (rxData[0] >> MCP4725_READ_RDY_Pos) & MCP4725_READ_RDY_Mask;
```

Comment:
The `MCP4725_READ_RDY_Mask` is not defined as I would usually define it. Usually the mask is applied to the register *before* shifting. In this case `MCP4725_READ_RDY_MASK` should be defined as `(1U << MCP4725_READ_RDY_POS)`, leading to the more simple statement:
```C
mcp4725_ready = rx_data[0] & MCP4725_READ_RDY_MASK;
```

Reply:
Change made